File: android/app/src/main/kotlin/app/transkey/mobile/PinyinHanziConverter.kt
package app.transkey.mobile

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import java.io.File

/**
 * Offline pinyin -> hanzi conversion backed by a compact SQLite dictionary
 * (pinyin -> hanzi, freq). Pinyin + hanzi come from CC-CEDICT (CC-BY-SA - keep
 * attribution); ranking frequency from jieba (MIT). Keys are toneless joined
 * pinyin (你好 -> "nihao"), so a full word reading is a direct lookup.
 *
 * Candidates = exact-pinyin matches, then predictive prefix completions, then a
 * greedy longest-prefix segmentation - all ranked by frequency. A downloaded
 * larger pack (filesDir/zhdict_large.db) supersedes the bundled one.
 */
class PinyinHanziConverter(private val context: Context) {

    @Volatile private var db: SQLiteDatabase? = null
    @Volatile private var syllables: Set<String> = emptySet()
    val isReady: Boolean get() = db != null

    companion object {
        private const val ASSET = "zhdict.db"
        private const val VER = "2"               // syllable table added
        private const val LARGE = "zhdict_large.db"
        private const val MAX_SYL_LEN = 6         // longest pinyin syllable (e.g. zhuang)
        private const val MAX_WORD = 6            // longest word (syllables) to group
    }

    fun ensureLoaded() {
        if (db != null) return
        synchronized(this) {
            if (db != null) return
            val large = File(context.filesDir, LARGE)
            val target = if (large.exists()) large else copyBundled()
            db = runCatching {
                SQLiteDatabase.openDatabase(target.path, null, SQLiteDatabase.OPEN_READONLY)
            }.getOrNull()
            syllables = query("SELECT s FROM syllables", emptyArray()).toHashSet()
        }
    }

    private fun copyBundled(): File {
        val out = File(context.filesDir, ASSET)
        val ver = File(context.filesDir, "zhdict.ver")
        if (out.exists() && ver.exists() && ver.readText() == VER) return out
        runCatching {
            context.assets.open(ASSET).use { input -> out.outputStream().use { input.copyTo(it) } }
            ver.writeText(VER)
        }
        return out
    }

    private fun query(sql: String, args: Array<String>): List<String> {
        val d = db ?: return emptyList()
        val out = ArrayList<String>()
        runCatching {
            d.rawQuery(sql, args).use { c -> while (c.moveToNext()) out.add(c.getString(0)) }
        }
        return out
    }

    private fun exact(pinyin: String, limit: Int) =
        query("SELECT hanzi FROM zhdict WHERE pinyin=? ORDER BY freq DESC LIMIT ?",
            arrayOf(pinyin, limit.toString()))

    /** Words whose pinyin starts with [pinyin] (predictive), excluding exact. */
    private fun prefix(pinyin: String, limit: Int) =
        query(
            "SELECT hanzi FROM zhdict WHERE pinyin>? AND pinyin<? ORDER BY freq DESC LIMIT ?",
            arrayOf(pinyin, pinyin + "{", limit.toString()), // '{' = 'z'+1
        )

    /** Split pinyin into valid syllables (greedy longest), or null if some
     *  tail can't be covered - e.g. an incomplete syllable while still typing. */
    private fun splitSyllables(pinyin: String): List<String>? {
        val out = ArrayList<String>()
        var i = 0
        while (i < pinyin.length) {
            var matched = false
            var j = minOf(i + MAX_SYL_LEN, pinyin.length)
            while (j > i) {
                val sub = pinyin.substring(i, j)
                if (sub in syllables) { out.add(sub); i = j; matched = true; break }
                j--
            }
            if (!matched) return null
        }
        return out
    }

    /**
     * Convert a full multi-syllable reading: split into syllables, then greedily
     * group the longest run of syllables that is a dictionary word, else take the
     * best single character per syllable (你很好 -> 你/很/好). Null if the input
     * isn't cleanly splittable into syllables (still mid-typing).
     */
    private fun segment(pinyin: String): String? {
        val syl = splitSyllables(pinyin) ?: return null
        if (syl.size < 2) return null // single syllable already covered by exact()
        val sb = StringBuilder()
        var i = 0
        while (i < syl.size) {
            var matched = false
            var j = minOf(i + MAX_WORD, syl.size)
            while (j > i) {
                val key = syl.subList(i, j).joinToString("")
                val s = exact(key, 1)
                if (s.isNotEmpty()) { sb.append(s[0]); i = j; matched = true; break }
                j--
            }
            if (!matched) {
                val s = exact(syl[i], 1)
                sb.append(if (s.isNotEmpty()) s[0] else syl[i]); i++
            }
        }
        return sb.toString()
    }

    /** Candidates for [pinyin] (lowercase a-z). Exact, prefix, then segmented. */
    fun convert(pinyin: String): List<String> {
        if (pinyin.isEmpty()) return emptyList()
        if (db == null) ensureLoaded()
        val res = LinkedHashSet<String>()
        // More than fit on the strip: the surplus is browsable in the expand-all
        // grid (▼), and number-key / long-press selection picks the first few.
        for (s in exact(pinyin, 9)) res.add(s)
        for (s in prefix(pinyin, 18)) res.add(s)
        segment(pinyin)?.let { if (it.isNotEmpty() && it != pinyin) res.add(it) }
        return res.toList()
    }
}

File: ios/TransKeyKeyboard/PinyinHanziConverter.swift
import Foundation

/// Offline pinyin -> hanzi conversion backed by a compact SQLite dictionary
/// (pinyin -> hanzi, freq). Pinyin + hanzi come from CC-CEDICT (CC-BY-SA - keep
/// attribution); ranking frequency from jieba (MIT). Keys are toneless joined
/// pinyin (你好 -> "nihao"), so a full word reading is a direct lookup.
///
/// Candidates = exact-pinyin matches, then predictive prefix completions, then
/// a greedy longest-prefix segmentation - all ranked by frequency. Port of the
/// Android PinyinHanziConverter (keep the two in sync).
final class PinyinHanziConverter {

    private static let maxSylLen = 6   // longest pinyin syllable (e.g. zhuang)
    private static let maxWord = 6     // longest word (syllables) to group

    private var dict: CJKDictionary?
    private var syllables: Set<String> = []
    var isReady: Bool { dict?.isReady == true }

    func ensureLoaded() {
        guard dict == nil else { return }
        let d = CJKDictionary(resource: "zhdict")
        syllables = Set(d.strings("SELECT s FROM syllables", []))
        dict = d
    }

    private func exact(_ pinyin: String, _ limit: Int) -> [String] {
        dict?.strings(
            "SELECT hanzi FROM zhdict WHERE pinyin=? ORDER BY freq DESC LIMIT \(limit)",
            [pinyin]) ?? []
    }

    /// Words whose pinyin starts with [pinyin] (predictive), excluding exact.
    private func prefix(_ pinyin: String, _ limit: Int) -> [String] {
        dict?.strings(
            "SELECT hanzi FROM zhdict WHERE pinyin>? AND pinyin<? ORDER BY freq DESC LIMIT \(limit)",
            [pinyin, pinyin + "{"]) ?? []   // '{' = 'z'+1
    }

    /// Split pinyin into valid syllables (greedy longest), or nil if some
    /// tail can't be covered - e.g. an incomplete syllable while still typing.
    private func splitSyllables(_ pinyin: String) -> [String]? {
        let chars = Array(pinyin)
        var out: [String] = []
        var i = 0
        while i < chars.count {
            var matched = false
            var j = min(i + Self.maxSylLen, chars.count)
            while j > i {
                let sub = String(chars[i..<j])
                if syllables.contains(sub) {
                    out.append(sub); i = j; matched = true; break
                }
                j -= 1
            }
            if !matched { return nil }
        }
        return out
    }

    /// Convert a full multi-syllable reading: split into syllables, then
    /// greedily group the longest run of syllables that is a dictionary word,
    /// else take the best single character per syllable (你很好 -> 你/很/好).
    /// Nil if the input isn't cleanly splittable (still mid-typing).
    private func segment(_ pinyin: String) -> String? {
        guard let syl = splitSyllables(pinyin), syl.count >= 2 else { return nil }
        var sb = ""
        var i = 0
        while i < syl.count {
            var matched = false
            var j = min(i + Self.maxWord, syl.count)
            while j > i {
                let key = syl[i..<j].joined()
                let s = exact(key, 1)
                if !s.isEmpty { sb += s[0]; i = j; matched = true; break }
                j -= 1
            }
            if !matched {
                let s = exact(syl[i], 1)
                sb += s.isEmpty ? syl[i] : s[0]
                i += 1
            }
        }
        return sb
    }

    /// Candidates for [pinyin] (lowercase a-z). Exact, prefix, then segmented.
    func convert(_ pinyin: String) -> [String] {
        guard !pinyin.isEmpty else { return [] }
        if dict == nil { ensureLoaded() }
        var seen = Set<String>()
        var res: [String] = []
        func add(_ s: String) {
            if seen.insert(s).inserted { res.append(s) }
        }
        for s in exact(pinyin, 9) { add(s) }
        for s in prefix(pinyin, 18) { add(s) }
        if let seg = segment(pinyin), !seg.isEmpty, seg != pinyin { add(seg) }
        return res
    }
}
